cpp/bindings/constraints.cpp

```diff
@@ -0,0 +1,109 @@
+      .export_values();
+
+  // Abstract class must be binded for derived classes to work
+  py::class_<LinearConstraint, LinearConstraintPtr>(m, "LinearConstraint")
+      .def_property_readonly("nbConstraints", &LinearConstraint::nbConstraints)
+      .def_property_readonly("nbVariables", &LinearConstraint::nbVariables)
+      .def_property_readonly("hasLinearInequalities",
+                             &LinearConstraint::hasLinearInequalities)
+      .def_property_readonly("hasUbounds", &LinearConstraint::hasUbounds)
+      .def_property_readonly("hasXbounds", &LinearConstraint::hasXbounds)
+      .def_property("discretizationType",
+                    (DiscretizationType(LinearConstraint::*)() const) &
+                        LinearConstraint::discretizationType,
+                    (void (LinearConstraint::*)(DiscretizationType)) &
+                        LinearConstraint::discretizationType)
+
+      .def ("computeParams", [](LinearConstraint& constraint, const GeometricPath& path, const Vector& gridpoints) -> py::tuple {
+            Vectors a, b, c, g;
+            Matrices F;
+            Bounds ub, xb;
+            constraint.computeParams(path, gridpoints, a, b, c, F, g, ub, xb);
+            py::list list;
+            if (constraint.hasLinearInequalities()) {
+              list.append(a);
+              list.append(b);
+              list.append(c);
+              list.append(F);
+              list.append(g);
+            } else
+              for (int i = 0; i < 5; ++i) list.append(NULL);
+            if (constraint.hasUbounds()) list.append(ub);
+            else list.append(NULL);
+            if (constraint.hasXbounds()) list.append(xb);
+            else list.append(NULL);
+            return list;
+          }
+        );
+    ;
+  py::class_<constraint::LinearJointVelocity,
+    std::shared_ptr<constraint::LinearJointVelocity>,
+    LinearConstraint>(
+      m, "LinearJointVelocity")
+      .def(py::init<const Vector &, const Vector &>())
+      ;
+  py::class_<constraint::LinearJointAcceleration,
+    std::shared_ptr<constraint::LinearJointAcceleration>,
+    LinearConstraint>(
+      m, "LinearJointAcceleration")
+      .def(py::init<const Vector &, const Vector &>())
+      ;
+
+  py::class_<constraint::JointTorque,
+    std::shared_ptr<constraint::JointTorque>,
+    LinearConstraint>(
+      m, "JointTorque");
+
+  {
+    auto mod_jointTorque = m.def_submodule("jointTorque");
+#ifdef BUILD_WITH_PINOCCHIO
+    py::module::import("pinocchio");
+    py::class_<constraint::jointTorque::Pinocchio<>,
+      std::shared_ptr<constraint::jointTorque::Pinocchio<> >,
+      constraint::JointTorque>(
+        mod_jointTorque, "Pinocchio")
+#ifdef PINOCCHIO_WITH_PYTHON_INTERFACE
+        .def(py::init([](py::object model, const toppra::Vector& fc) {
+              return toppra::constraint::jointTorque::Pinocchio<>(
+                  boost::python::extract<const pinocchio::Model&>(model.ptr()), fc);
```

Comment:
The only potential bug is here. The Pinocchio Model is held by Python. If Python deletes the object, the joint torque constraint will point to a deleted model.

Reply:
model will be a copy and is owned by pybind, no? In any case it will go out-of-scope after the constructor terminates and the pointer will be invalid all the same.

Reply:
Sorry for the confusion:
by model, I meant the `pInocchio::Model` object, not the `py::object`. The `py::object`, when created, increases the reference to the already existing Python object wrapping the `pinocchio::Model`. This `py::object` is temporary, so the owning is temporary too. After this method returns, the only person who owns the object is Python.

This means that
```python
def function():
  model = createModel()
  return toppra.cpp.jointTorque.Pinocchio(model, np.zeros(model.nv))

c = function()
```
will segv (or have undefined behavior) when using `c`.

Reply:
I think pybind11 offers a way of binding the lifetime of a Python object to another object. This could fix the issue.

https://pybind11.readthedocs.io/en/master/advanced/functions.html#keep-alive

Reply:
This is now fixed.